src/flat.mjs

```diff
@@ -1,29 +1,39 @@
 import ensureIterable from './internal/ensure-iterable'
 
-function * flat (depth, iterable) {
-  if (depth === 0) {
-    yield * ensureIterable(iterable)
-  } else {
-    for (const iter of ensureIterable(iterable)) {
-      if (typeof iter !== 'string' && typeof iter[Symbol.iterator] === 'function') {
-        yield * flat(depth - 1, iter)
-      } else {
-        yield iter
+const defaultShouldIFlat = (depth) => {
+  if (typeof depth === 'function') {
+    return depth
+  }
+  if (typeof depth === 'number') {
+    return (currentDepth, iter) =>
+      currentDepth <= depth && typeof iter[Symbol.iterator] === 'function' && typeof iter !== 'string'
+  }
+  throw new Error('flat: "depth" can be a function or a number')
+}
+
+function flat (shouldIFlat, iterable) {
+  function * _flat (currentDepth, iterable) {
+    if (shouldIFlat(currentDepth, iterable)) {
+      for (const iter of iterable) {
+        yield * _flat(currentDepth + 1, iter)
       }
+    } else {
+      yield iterable
     }
   }
+  return _flat(0, ensureIterable(iterable))
 }
 
-export default function curriedFlat (depth, iterable) {
+export default function curriedFlat () {
```

Comment:
I see you're using `arguments` variable which was never recommended in JavaScript.
A better alternative would be rest parameters.

```javascript
export default function curriedFlat (...args)
```

Reply:
Rest args are nicer and more readable. Ok